notebooks/02-ingestion/05-partitioning-data.ipynb

```diff
@@ -0,0 +1,717 @@
+    "\n",
+    "display.sql(sql.format(\"flights_hour\"))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d83c7eb1-6d0a-4a08-aa1b-03ec8b755539",
+   "metadata": {},
+   "source": [
+    "### Performance Results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "dc4c2524-0eba-4bc6-84ed-da3a25aa5fbe",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# run 100 queries with each table\n",
+    "use_table=\"flights_hour\"\n",
+    "print (f\"{use_table} {measure_query(sql.format(use_table), 100)}\")\n",
+    "\n",
+    "use_table=\"flights_year\"\n",
+    "print (f\"{use_table} {measure_query(sql.format(use_table), 100)}\")\n",
+    "\n",
+    "use_table=\"flights_year_50k\"\n",
+    "print (f\"{use_table} {measure_query(sql.format(use_table), 100)}\")\n",
+    "\n",
+    "use_table=\"flights_year_IATA\"\n",
+    "print (f\"{use_table} {measure_query(sql.format(use_table), 100)}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d0716364-af1f-4eb1-9e75-3ce3c5617344",
+   "metadata": {},
+   "source": [
+    "### Single Concurrency Results\n",
+    "\n",
+    "Given that the characteristics of your laptop or other execution environment are likely different than the `Apple M2 Pro` that this was built on, your results will vary, my results were:\n",
+    "```\n",
+    "flights_hour Results      = avg:28.848 ms   min:25.198 ms   max:64.379 ms\n",
+    "flights_year Results      = avg:15.451 ms   min:13.891 ms   max:17.940 ms\n",
+    "flights_year_50k Results  = avg:12.844 ms   min:11.594 ms   max:15.130 ms\n",
+    "flights_year_IATA Results = avg:12.574 ms   min:11.221 ms   max:15.204 ms\n",
+    "```\n",
+    "\n",
+    "- `flights_hour` - With segment granularity of hour, there are approximately 30*24= 720 segment files to scan, there is no pruning, so even though the segments are tiny, each one is examined separately and the results are then merged, so doing that 720 times makes this the slowest option.\n",
+    "- `flights_year` - Here we are using a single segment, it is much larger at 500k+ rows, but Druid can still use the indexes associated to all dimension columns to process this pretty fast and it only does it once, so this is faster then hourly. This shows how the number of segments needed for a query is a factor for performance. Less segments to process tends to improve perfomance, but not always.\n",
+    "- `flights_year_50k` - Now each query had to process twelve 50k segments, but given that the historical does this concurrently on multiple threads, the parallelism is enough to improve over the single 500k segment of the `flights_year` table. \n",
+    "- `flights_year_IATA` - Given the shard spec of the 12 resulting segments, we can see that there are rows with IATA Code \"AA\" in 2 of the segments. They are both processed in parallel and there are less segments to process, so this results in the fastest. \n",
+    "\n",
+    "### Effects of Higher Concurrency\n",
```

Comment:
Hmmmmmm so I'm thinking - you've proved what PARTITIONED BY and CLUSTERED BY and maxrows does, and you've shown that you need to think about the performance.

The rest of this notebook is arguably an opportunity to have a deeper dive. When I did this notebook I got to this point and felt I'd learned enough to know what these two things did, and that it impacted query performance.

Maybe all that follows could be spun off into a blog post?

Reply:
Concurrency is an important consideration when deciding how to resolve a set of query patterns. The point here is that you can see the effects of better partitioning on query performance stability as concurrency grows.

I like it as the final element in proving the design strategy.